sql: TestSchemaChangePurgeFailure failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/62c976dafc7f48f0a52142310a8fdc42908c30a5

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stressrace instead of stress and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stress TESTS=TestSchemaChangePurgeFailure PKG=github.com/cockroachdb/cockroach/pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-stderr=false -maxtime 20m -timeout 10m'
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=961469&tab=buildLog

```
I181012 06:51:34.845802 107179 sql/event_log.go:126  [n1,client=127.0.0.1:41898,user=root] Event: "create_table", target: 53, info: {TableName:t.public.test Statement:CREATE TABLE t.public.test (k INT PRIMARY KEY, v INT) User:root}
I181012 06:51:34.860952 107197 storage/replica_command.go:298  [n1,split,s1,r20/1:/{Table/23-Max}] initiating a split of this range at key /Table/53 [r21]
I181012 06:51:35.148326 107179 sql/lease.go:345  [n1,client=127.0.0.1:41898,user=root] publish: descID=53 (test) version=2 mtime=2018-10-12 06:51:35.093760965 +0000 UTC
I181012 06:51:35.243334 107179 sql/event_log.go:126  [n1,client=127.0.0.1:41898,user=root] Event: "create_index", target: 53, info: {TableName:t.public.test IndexName:foo Statement:CREATE UNIQUE INDEX foo ON t.public.test (v) User:root MutationID:1}
I181012 06:51:35.312095 107267 sql/lease.go:877  new lease: 53("test") ver=2:1539327465.110907697,0, refcount=0
I181012 06:51:35.485167 106949 sql/lease.go:345  [n1,scExec] publish: descID=53 (test) version=3 mtime=2018-10-12 06:51:35.483388781 +0000 UTC
I181012 06:51:35.551745 106949 sql/lease.go:315  publish (1 leases): desc=[{test 53 2}]
I181012 06:51:35.555253 107165 sql/lease.go:877  new lease: 53("test") ver=3:1539327407.976041060,0, refcount=0
I181012 06:51:35.609356 106949 sql/lease.go:315  publish (1 leases): desc=[{test 53 2}]
I181012 06:51:35.712915 106949 sql/backfill.go:131  [n1,scExec] Running backfill for "test", v=3, m=1
W181012 06:51:35.828707 106949 sql/schema_changer.go:866  [n1,scExec] reversing schema change 390780338523144193 due to irrecoverable error: duplicate key value (v)=(201) violates unique constraint "foo"
W181012 06:51:35.864763 106949 sql/schema_changer.go:1146  [n1,scExec] reverse schema change mutation: {Descriptor_:0xc421054470 State:DELETE_AND_WRITE_ONLY Direction:ADD MutationID:1 Rollback:false}
I181012 06:51:35.865114 106949 sql/lease.go:345  [n1,scExec] publish: descID=53 (test) version=4 mtime=2018-10-12 06:51:35.86289029 +0000 UTC
I181012 06:51:36.070605 106949 sql/event_log.go:126  [n1,scExec] Event: "reverse_schema_change", target: 53, info: {Error:duplicate key value (v)=(201) violates unique constraint "foo" MutationID:1}
I181012 06:51:36.117989 107273 sql/lease.go:877  new lease: 53("test") ver=4:1539327393.798664448,0, refcount=0
I181012 06:51:36.220470 106949 sql/lease.go:345  [n1,scExec] publish: descID=53 (test) version=5 mtime=2018-10-12 06:51:36.215318642 +0000 UTC
I181012 06:51:36.275277 106949 sql/lease.go:315  publish (1 leases): desc=[{test 53 4}]
I181012 06:51:36.311751 107229 sql/lease.go:877  new lease: 53("test") ver=5:1539327433.787755650,0, refcount=0
I181012 06:51:36.344345 106949 sql/lease.go:315  publish (1 leases): desc=[{test 53 4}]
I181012 06:51:36.428499 106949 sql/backfill.go:131  [n1,scExec] Running backfill for "test", v=5, m=1
I181012 06:51:36.452962 106949 sql/lease.go:345  [n1,scExec] publish: descID=53 (test) version=6 mtime=2018-10-12 06:51:36.451244816 +0000 UTC
I181012 06:51:36.556242 106949 sql/event_log.go:126  [n1,scExec] Event: "finish_schema_change_rollback", target: 53, info: {MutationID:1}
I181012 06:51:36.597824 106949 sql/lease.go:315  publish (1 leases): desc=[{test 53 5}]
I181012 06:51:36.610053 107352 sql/lease.go:877  new lease: 53("test") ver=6:1539327339.303763310,0, refcount=0
I181012 06:51:36.676383 106949 sql/lease.go:315  publish (1 leases): desc=[{test 53 5}]
W181012 06:51:36.758491 106949 sql/schema_changer.go:1550  [n1] Error executing schema change: duplicate key value (v)=(201) violates unique constraint "foo"
I181012 06:51:36.911190 106409 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] transport racer
1      [async] closedts-rangefeed-subscriber
I181012 06:51:36.916812 106409 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] transport racer
I181012 06:51:36.974684 106682 kv/transport_race.go:113  transport race promotion: ran 45 iterations on up to 1395 requests
--- FAIL: TestSchemaChangePurgeFailure (5.87s)
	schema_changer_test.go:1541: <nil>

```

--- a/pkg/sql/schema_changer_test.go
+++ b/pkg/sql/schema_changer_test.go
@@ -1474,7 +1474,7 @@ func TestSchemaChangePurgeFailure(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	params, _ := tests.CreateTestServerParams()
 	const chunkSize = 200
-	var enableAsyncSchemaChanges uint32 = 1
+	var enableAsyncSchemaChanges uint32
 	var attempts int32
 	// attempt 1: write the first chunk of the index.
 	// attempt 2: write the second chunk and hit a unique constraint
@@ -1541,6 +1541,15 @@ CREATE TABLE t.test (k INT PRIMARY KEY, v INT);
 		t.Fatal(err)
 	}
 
+	// The index doesn't exist
+	if _, err := sqlDB.Query(
+		`SELECT v from t.test@foo`,
+	); !testutils.IsError(err, "index .* not found") {
+		t.Fatal(err)
+	}
+
+	// Allow async schema change purge to attempt backfill and error.
+	atomic.StoreUint32(&enableAsyncSchemaChanges, 1)
 	tableDesc := sqlbase.GetTableDescriptor(kvDB, "t", "test")
 	if _, err := addImmediateGCZoneConfig(sqlDB, tableDesc.ID); err != nil {
 		t.Fatal(err)
@@ -1554,13 +1563,6 @@ CREATE TABLE t.test (k INT PRIMARY KEY, v INT);
 		}
 		return nil
 	})
-
-	// The index doesn't exist
-	if _, err := sqlDB.Query(
-		`SELECT v from t.test@foo`,
-	); !testutils.IsError(err, "index .* not found") {
-		t.Fatal(err)
-	}
 
 	tableDesc = sqlbase.GetTableDescriptor(kvDB, "t", "test")
 

